smallsetup/StartK8s.ps1

```diff
@@ -312,67 +383,30 @@ if (!$WSL -and !$isReusingExistingLinuxComputer) {
-    $i = 0;
-    while ($true) {
-        $i++
-        Write-Log "VM Handling loop (iteration #$i):"
-        Start-Sleep -s 4
-
-        if ( $i -eq 1 ) {
-            Write-Log "           stopping VM ($i)"
-            Stop-VM -Name $controlPlaneVMHostName -Force -WarningAction SilentlyContinue
-
-            $state = (Get-VM -Name $controlPlaneVMHostName).State -eq [Microsoft.HyperV.PowerShell.VMState]::Off
-            while (!$state) {
-                Write-Log "           still waiting for stop, current VM state: $(Get-VM -Name $controlPlaneVMHostName | Select-Object -expand 'State')"
-                Start-Sleep -s 1
-            }
-
-            Write-Log "           re-starting VM ($i)"
-            Start-VM -Name $controlPlaneVMHostName
-            Start-Sleep -s 4
-        }
-
-        $con = Test-Connection $ipControlPlane -Count 1 -ErrorAction SilentlyContinue
-        if ($con) {
-            Write-Log "           ping succeeded to $controlPlaneVMHostName VM"
-            $startStatus = (Get-VM -Name $controlPlaneVMHostName).State -eq [Microsoft.HyperV.PowerShell.VMState]::Running
-            if ($startStatus) {
-                Write-Log "           $controlPlaneVMHostName VM Started"
-                break;
-            }
-        }
+# VM restart loop
+$i = 0
+while($true) {
+    $i++
+    Restart-ControlPlane -ControlPlaneVMName $controlPlaneVMHostName `
+                     -ControlPlaneIpAddr $ipControlPlane `
+                     -WSL $WSL `
+                     -ReusingExistingLinuxMachine $isReusingExistingLinuxComputer
+    $controlPlaneCni0IpAddr = Get-Cni0IpAddressInControlPlaneUsingSshWithRetries -Retries 10 -RetryTimeoutInSeconds 5
+    $expectedControlPlaneCni0IpAddr = Get-ConfiguredMasterNetworkInterfaceCni0IP
+                 
+    if ($controlPlaneCni0IpAddr -ne $expectedControlPlaneCni0IpAddr) {
+        Write-Log "cni0 interface in $controlPlaneVMHostName is not correctly initialized."
+        Write-Log "           Expected:$expectedControlPlaneCni0IpAddr"
+        Write-Log "           Actual:$controlPlaneCni0IpAddr"
 
         if ($i -eq 3) {
-            # If the connection did not succeed or VM is not in Running state, try starting again for three times
-            $startCycle = 0
-            $startState = (Get-VM -Name $controlPlaneVMHostName).State -eq [Microsoft.HyperV.PowerShell.VMState]::Running
-            while (!$startState -And $startCycle -lt 3) {
-                $startCycle++
-                Write-Log "           still waiting for start, current VM state: $(Get-VM -Name $controlPlaneVMHostName | Select-Object -expand 'State')"
-                Start-VM -Name $controlPlaneVMHostName
-                Start-Sleep -s 4
-                $startState = (Get-VM -Name $controlPlaneVMHostName).State -eq [Microsoft.HyperV.PowerShell.VMState]::Running
-            }
+            throw "cni0 interface in $controlPlaneVMHostName is not correctly initialized after $i retries."
```

Comment:
Before throwing exception, is there a workaround which we can inform user to try like start and stop? Not mandatory, just lay out a set of instructions to user if there a way to overcome this problem.

Reply:
Ideally, the restart of the kubemaster should solve the incorrectly initialized cni0. But if it is not the case, then we have to look as to why that is the case.

Currently, when the exception is thrown at this stage, all K2s services are not deployed. This means that we cannot connect to kubemaster via system ssh commands because the command would first check if k2s is running.